AGENTS.md

```diff
@@ -0,0 +1,187 @@
+- `InitializingPayload`: First-time deployment, fast progress with tolerance for transient errors
+- Release payloads stored in two directories:
+  - `manifests/`: CVO's own manifests
+  - `release-manifests/`: Manifests for other cluster operators
+- Task graph orchestrates ordered application of manifests based on dependencies
+
+#### Resource Libraries (lib/)
+- `resourcebuilder/`: Builds Kubernetes resources from manifests
+- `resourceapply/`: Applies resources with proper merge semantics for different API types
+- `resourcemerge/`: Merges resource specifications (handles fields like `hostUsers` flag)
+- `resourceread/`: Reads and validates manifests
+- `capability/`: Manages cluster capability filtering
+
+#### Update Service Integration (pkg/cincinnati/)
+- Communicates with OpenShift Update Service (Cincinnati) to fetch available updates
+- Parses update graphs with nodes (release versions) and edges (upgrade paths)
+- Supports conditional edges based on cluster conditions
+
+#### Preconditions (pkg/payload/precondition/)
+- Validates cluster readiness before applying updates
+- Checks prevent upgrades when cluster is in unhealthy state
+- ClusterVersion-specific preconditions in `precondition/clusterversion/`
+
+### Entry Points
+
+#### cmd/cluster-version-operator/main.go
+- Uses cobra for CLI structure
+- Actual start logic delegated to `pkg/start/` package
+
+#### pkg/start/start.go
+- Initializes and launches core CVO control loops
+- Sets up client connections, informers, and signal handling
+- Creates and starts the main `Operator` instance
+
+### Key Patterns
+
+#### Release Image Structure
+- Release images are OCI container images containing:
+  - CVO binary at a specific version
+  - Manifest files in `manifests/` and `release-manifests/`
+  - `release-metadata` file with Cincinnati update graph info
+  - `image-references` file mapping component names to images
+
+#### Update Flow
+1. CVO fetches available updates from configured upstream (OSUS)
+2. Updates stored in `status.availableUpdates` of `ClusterVersion` resource
+3. User/admin sets `spec.desiredUpdate` to trigger upgrade
+4. CVO retrieves new release payload image
+5. Validates payload signatures (if signature verification enabled)
+6. Applies manifests systematically using task graph
+7. Monitors cluster operator readiness during rollout
+8. Updates `ClusterVersion` status conditions throughout process
+
+#### Manifest Application Order
+- Task graph in `pkg/payload/task_graph.go` determines ordering
+- Dependencies between operators enforced during updates
+- Parallel application during reconciliation for faster recovery
+
+## Development Workflows
+
+### Testing changes locally
+```bash
+# Build custom release payload with development CVO
+oc adm release new \
+  --from-release=quay.io/openshift-release-dev/ocp-release:${TAG} \
+  --to-image-base=${REPO}/origin-cluster-version-operator:latest \
+  --to-image=${REPO}/ocp-release:latest
```

Comment:
I have gone at least 4 years without having to do this.  Possibly thanks to Cluster Bot?  Not clear to me if we need to cover it anymore, or if we do need to cover it, who we expect to use it when?  Maybe I'm unique, and other update devs are using this?